File: evaluators/trajectory/subset.py
from evaluators.types import ChatCompletionMessage, EvaluatorResult
from evaluators.utils import _run_evaluator
from evaluators.trajectory.utils import _is_trajectory_superset
from typing import Any


def trajectory_subset(
    *,
    outputs: list[ChatCompletionMessage],
    reference_outputs: list[ChatCompletionMessage],
    **kwargs: Any,
) -> EvaluatorResult:
    """
    Evaluate whether an agent trajectory and called tools is a subset of a reference trajectory and called tools.
    This means the agent called a subset of the tools specified in the reference trajectory.

    Args:
        outputs (list[ChatCompletionMessage]): Actual trajectory the agent followed
        reference_outputs (list[ChatCompletionMessage]): Ideal reference trajectory the agent should have followed

    Returns:
        EvaluatorResult: Contains a score of 1.0 if trajectory (including called tools) matches, 0.0 otherwise
    """

    def get_score():
        if outputs is None or reference_outputs is None:
            raise ValueError(
                "Trajectory subset match requires both outputs and reference_outputs"
            )
        is_superset = _is_trajectory_superset(reference_outputs, outputs)
        return 1.0 if is_superset else 0.0

    return _run_evaluator(
        run_name="trajectory_subset",
        evaluator_fn=get_score,
        feedback_key="trajectory_subset",
    )

File: js/src/evaluators/trajectory/subset.ts
import { BaseMessage } from "@langchain/core/messages";
import { ChatCompletionMessage, EvaluatorResult } from "../types.js";
import { _normalizeToOpenAIMessagesList, _runEvaluator } from "../utils.js";
import { _isTrajectorySuperset } from "./utils.js";

/**
 * Evaluate whether an agent trajectory and called tools is a subset of a reference trajectory and called tools.
 * This means the agent called a subset of the tools specified in the reference trajectory.
 *
 * @param params - The parameters for trajectory subset evaluation
 * @param params.outputs - Actual trajectory the agent followed.
 *    May be a list of OpenAI messages, a list of LangChain messages, or a dictionary containing
 *    a "messages" key with one of the above.
 * @param params.reference_outputs - Ideal reference trajectory the agent should have followed.
 *    May be a list of OpenAI messages, a list of LangChain messages, or a dictionary containing
 *    a "messages" key with one of the above.
 * @returns EvaluatorResult containing a score of true if trajectory (including called tools) matches, false otherwise
 */
export async function trajectorySubset(params: {
  outputs:
    | ChatCompletionMessage[]
    | BaseMessage[]
    | { messages: (BaseMessage | ChatCompletionMessage)[] };
  referenceOutputs:
    | ChatCompletionMessage[]
    | BaseMessage[]
    | { messages: (BaseMessage | ChatCompletionMessage)[] };
}): Promise<EvaluatorResult> {
  const { outputs, referenceOutputs } = params;
  const outputsList = _normalizeToOpenAIMessagesList(outputs);
  const referenceOutputsList = _normalizeToOpenAIMessagesList(referenceOutputs);

  const getScore = async () => {
    if (outputsList == null || referenceOutputsList == null) {
      throw new Error(
        "Trajectory subset match requires both outputs and reference_outputs"
      );
    }
    const isSubset = _isTrajectorySuperset(referenceOutputsList, outputsList);
    return isSubset;
  };
  return _runEvaluator(
    "trajectory_subset",
    getScore,
    "trajectory_subset",
    params
  );
}
